### PR TITLE
llvmPackages_15.libcxx: specify `LIBCXX_CXX_ABI_{LIBRARY_PATH,INCLUDE_PATHS}` for all cxxabis

### LINKED ISSUE #214524: llvmPackages_15: Building with `clang++` using `libcxx`/`-stdlib=libc++` fails
### Describe the bug
Using `-lc++` with `clang++` seems to be broken with `llvmPackages_15.clang` and `llvmPackages_15.libcxxClang`. It fails with `ld: cannot find l: No such file or directory`. https://github.com/NixOS/nixpkgs/issues/213144 also notes this.

### Steps To Reproduce

Running `clang++ -stdlib=libc++ main.cpp` yields `-/nix/store/r2b9k28c6aghczpqfvh71y9xavm7rr68-binutils-2.39/bin/ld: cannot find l: No such file or directory`. The `main.cpp` contains just an empty `main()` and I am using this flake's devShell:
```nix
{
  description = "Test";

  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/nixos-unstable";
  };

  outputs = { self, nixpkgs }: let
    lib = nixpkgs.lib;
    pkgs = nixpkgs.legacyPackages.x86_64-linux;
    llvmPkgs = pkgs.llvmPackages_15;
  in {
    devShells.x86_64-linux.default = pkgs.mkShell {
      packages = [
        llvmPkgs.clang

        llvmPkgs.libcxx
        llvmPkgs.libcxxabi
      ];
    };
  };
}
```
The same also occurs when using a shell with only `llvmPkgs.libcxxClang` and when trying to use `libcxx` inside a nix build.

I am pretty sure that this is caused by `lib/libc++.so` in `llvmPackages_15.libcxx` containing `INPUT(libc++.so.1 -l)` instead of `INPUT(libc++.so.1 -lc++abi)` (which it does for `llvmPackages_14.libcxx`). I don't know yet, what the root cause for that is or how to fix that.

### Notify maintainers

@rrbutani @Ericson2314 @sternenseemann @lovek323 @dtzWill @primeos

(So it shows up in the tracking issue https://github.com/NixOS/nixpkgs/issues/213033)

### PR DESCRIPTION
**Pull request by rrbutani**
###### Description of changes

This PR mostly just fixes #214524 but also takes the time to:
  - update `tests.cc-wrapper` to test modern `gcc`/`clang` stdenvs (these tests catch the issue in #214524)
  - add some extra `libcxx` build flags that we were missing: https://github.com/NixOS/nixpkgs/commit/6828e2353b60bf1e1951ca4516a4ed735d162db8

TODO:
  - [ ]  test the LLVM FreeBSD cross stdenv (`cxxrt`)
  - [ ] make sure all the `tests.cc-wrapper.*` tests pass
  - [ ] modify `llvmPackages_git` as well
  - [ ] modify `llvmPackages_16` as is appropriate

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SchrodingerZhu** [comment]
Nothing urgent. I am just hoping to setup a LLVM15 environment with `libcxxStdenv`

**Comment by aaronmondal** [comment]
This is probably also required for LLVM 16.

**Comment by rrbutani** [comment]
Apologies, once again, for the delay; I should have time to finish this and the remaining issues in #213033 this weekend.

---

> This is probably also required for LLVM 16.

I believe this actually isn't; [as mentioned](https://github.com/NixOS/nixpkgs/issues/214524#issuecomment-1429146432) the patch we're applying here is present in LLVM 16.

We'll see though; the goal of the other half of this PR is to actually test the LLVM stdenvs a bit so we can catch issues like these 🤞.

**Comment by aaronmondal** [comment]
> I believe this actually isn't; https://github.com/NixOS/nixpkgs/issues/214524#issuecomment-1429146432

Oh I totally overlooked that. Sorry 🙏

> the goal of the other half of this PR is to actually test the LLVM stdenvs a bit so we can catch issues like these

Does the nixpkgs CI run global tests with the LLVM stdenv on x86_64-linux (with glibc, not with musl)? I keep running into curious issues that make it seem like the LLVM toolchain is not really tested against downstream projects. I also noticed that there is no build cache for `pkgsLLVM`.

Maybe it would make sense to dogfood the LLVM toolchain to a few buildsystem packages like CMake, Meson and Bazel? The tests for such packages might be a helpful sanity check.

**Comment by rrbutani** [comment]
> Oh I totally overlooked that. Sorry

no worries!

> Does the nixpkgs CI run global tests with the LLVM stdenv on x86_64-linux (with glibc, not with musl)? I keep running into curious issues that make it seem like the LLVM toolchain is not really tested against downstream projects. I also noticed that there is no build cache for `pkgsLLVM`.

I'm not the expert here but I believe the most testing the LLVM stdenv currently gets (outside of packages that explicitly use it in their builds) is [this](https://github.com/NixOS/nixpkgs/blob/3a2e5671843bd21ec71a8a29262a96b03bde2239/pkgs/top-level/release.nix#L218-L220) (builds the stdenv but does not use it). This PR adds the [`cc-wrapper` tests](https://github.com/NixOS/nixpkgs/blob/b2cf9511b3781b2e04bfb9d41be8c22a0ff9e5c5/pkgs/test/default.nix#L6).

> Maybe it would make sense to dogfood the LLVM toolchain to a few buildsystem packages like CMake, Meson and Bazel? The tests for such packages might be a helpful sanity check.

Definitely! `release-cross` is set up for [exactly this kind of testing](https://github.com/NixOS/nixpkgs/blob/b6e28d2a32e793a5091a871709c31e7f258d4a9d/pkgs/top-level/release-cross.nix#L24-L98) but at present (AFAIK) it does not test usage of the LLVM stdenv (i.e. `useLLVM = true` isn't in the cross system configs it tests). IIRC I don't think any channels actually block on that jobset (the health doesn't look great [either](https://hydra.nixos.org/jobset/nixpkgs/cross-trunk#tabs-evaluations)).

---

My experience matches yours: I've seen lots of breakage when trying to use the LLVM stdenvs, in their various forms; better testing seems undoubtedly valuable here.

At the moment though, I've been focusing on trying to deduplicate the LLVM package sets (which will make it a little easier to introduce tests and fix bugs and such).

**Comment by RaitoBezarius** [comment]
We should definitely extract the cc-wrapper tests stuff and merge it ASAP, for the rest, it can wait ofc, but I am available to work on reviewing or helping on that. :)

**Comment by centromere** [comment]
Any word on when this can be merged? I have a package requiring LLVM 15 which I am unable to build without this patch.

I also needed the change mentioned by @LogicalOverflow [here](https://github.com/NixOS/nixpkgs/issues/214524#issuecomment-1431745905).

**Comment by DieracDelta** [comment]
+1 would also love to see this merged

**Comment by centromere** [comment]
Hi @rrbutani. Did you have any success?

**Comment by Artturin** [comment]
Test changes split to https://github.com/NixOS/nixpkgs/pull/253752

File: pkgs/development/compilers/llvm/15/libcxx/default.nix [hunk 1]
```diff
@@ -47,6 +47,15 @@ stdenv.mkDerivation rec {
 
   patches = [
     ./gnu-install-dirs.patch
+    # See:
+    #   - https://reviews.llvm.org/D133566
+    #   - https://github.com/NixOS/nixpkgs/issues/214524#issuecomment-1429146432
+    # !!! Drop in LLVM 16+
+    (fetchpatch {
```
Comment (Artturin) [inline pkgs/development/compilers/llvm/15/libcxx/default.nix:54]: I've split this to https://github.com/NixOS/nixpkgs/pull/253936 as it's ready while the other issue still needs reviewing

**Comment by Artturin** [comment]
CC @pwaller who did many fixes https://github.com/NixOS/nixpkgs/pull/246577

**Comment by pwaller** [comment]
> CC @pwaller who did many fixes #246577

[only had a cursory look] Seems reasonable at a glance.

File: pkgs/development/compilers/llvm/15/libcxx/default.nix
Comment (Artturin) [inline pkgs/development/compilers/llvm/15/libcxx/default.nix:1]: are both of the changes needed or only one or is the libcxx/default.nix change unneeded or is it needed in 16 too

**Comment by Artturin** [comment]
https://github.com/NixOS/nixpkgs/pull/255488

**Comment by centromere** [comment]
What is preventing this from being merged?

**Comment by emilazy** [comment]
I understand this was fixed, I think by https://github.com/NixOS/nixpkgs/pull/307184. Sorry that this stalled out!

**Pull request closed without merging**